hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3ARemoteFileChanged.java

```diff
@@ -432,6 +433,42 @@ public void testReadFileChangedOutOfSyncMetadata() throws Throwable {
     }
   }
 
+  /**
+   * Verifies that when the openFile builder is passed in a status,
+   * then that is used to eliminate the getFileStatus call in open();
+   * thus the version and etag passed down are still used.
+   */
+  @Test
+  public void testOpenFileWithStatus() throws Throwable {
+    final Path testpath = path("testOpenFileWithStatus.dat");
+    final byte[] dataset = TEST_DATA_BYTES;
+    S3AFileStatus originalStatus =
+        writeFile(testpath, dataset, dataset.length, true);
+
+    // forge a file status with a different tag
+    S3AFileStatus forgedStatus =
+        S3AFileStatus.fromFileStatus(originalStatus, Tristate.FALSE,
+            originalStatus.getETag() + "-fake",
+            originalStatus.getVersionId() + "");
+    fs.getMetadataStore().put(
+        new PathMetadata(forgedStatus, Tristate.FALSE, false));
+
+    // By passing in the status open() doesn't need to check s3guard
+    // And hence the existing file is opened
+    try (FSDataInputStream instream = fs.openFile(testpath)
+        .withFileStatus(originalStatus)
+        .build().get()) {
+       instream.read();
+    }
+
+    // and this holds for S3A Located Status
+    try (FSDataInputStream instream = fs.openFile(testpath)
+        .withFileStatus(new S3ALocatedFileStatus(originalStatus, null))
+        .build().get()) {
+      instream.read();
+    }
+  }
```

Comment:
What about vanilla `fs.openFile(testPath)` without file status? It will simply fail right? Do we need a test for that?

Reply:
you mean with the file version changed? It shouldn't notice the version issues without s3guard, will fai on read() after a number of retriesl when s3guard version doesn't match that in the store

Reply:
The two try-with tests are testing that with a good FileStatus passed, it's able to skip the S3Guard and hence no errors are reported regardless of change detection policy.

I was thinking is it worth to make sure if we pass the wrong file status, it will error out because S3Guard is tested to have bad state. For e.g.
```
    try(FSDataInputStream instream = fs.openFile(testpath)
        .build().get()) {
      try {
        instream.read();
       // No exception only if we don't enforce change detection as exception
        assertTrue(changeDetectionMode.equals(CHANGE_DETECT_MODE_NONE) ||
            changeDetectionMode.equals(CHANGE_DETECT_MODE_WARN));
      } catch (Exception ignored) {
        // Ignored.
      }
    }
``` 

I think again and now guess this extra test perhaps is not required to demonstrate the withFileStatus is being honored.

Reply:
done. Also a contract test to verify that the status must be non-null